Redis Stream Inbound and Outbound Channel adapter documentation

rohanmukesh opened the pull request.

Commit:
Add Redis Stream Inbound and Outbound Channel adapter

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -791,12 +791,104 @@ You should set the expiry at a large enough value to prevent this condition, but
+    private FluxMessageChannel messageChannel;
+
+    public void sendMessage(Flux<Object> myFlux) {
+        messageChannel.subscribeTo(myFlux.map(myObj -> new GenericMessage<>(myObj)));
```
This is too complicated in a simple configuration sample.
Why do we need this piece of code at all?
Why just show plain `ReactiveRedisStreamMessageHandler` with that `@ServiceActivator` not enough?
Making everything else implied.

Anyway: why stick only with `FluxMessageChannel`?
We really can't use any other channels?

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -791,12 +791,104 @@ You should set the expiry at a large enough value to prevent this condition, but
+}
+----
+====
+
```
I wonder if we can add configuration options list like you did in the inbound section...

Especially people are going to raise concern how the payload is serialized into Redis record.
Although I see there is  an `extractPayload` option, which needs some details here to clarify...

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -791,12 +791,104 @@ You should set the expiry at a large enough value to prevent this condition, but
+    public IntegrationFlow myFlow() {
+        return IntegrationFlows
+                    .from(fluxMessageChannel())
+                    .get();
```
Again: this all looks too much just for explaining the component configuration.
It is really must not matter how I am going to use it. 
Why do you stick only with the `FluxMessageChannel`?
And why do you advertise an `IntegrationFlow` just with one channel?

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -791,12 +791,104 @@ You should set the expiry at a large enough value to prevent this condition, but
+    @Bean
+    ReactiveRedisStreamMessageProducer reactiveRedisStreamProducer(ReactiveRedisConnectionFactory reactiveRedisConnectionFactory) {
+    	ReactiveRedisStreamMessageProducer messageProducer =
+                new ReactiveRedisStreamMessageProducer(reactiveRedisConnectionFactory, "myStreamKey");
```
I think `ConnectionFactory` and `streamName` should be described below as well.

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -791,12 +791,104 @@ You should set the expiry at a large enough value to prevent this condition, but
+----
+<1> `StreamReceiver.StreamReceiverOptions` Options to consume redis stream using reactive infrastructure.
+<2> A `SmartLifecycle` attribute to specify whether this endpoint should start automatically after the application context start or not.
+It defaults to `true`. if `false`, RedisStreamMessageProducer should be started manually `messageProducer.start()`.
```
https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line

And looks like the next sentence doesn't start from capital letter...
The `RedisStreamMessageProducer` word must be wrapped into a code snippet.

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -791,12 +791,104 @@ You should set the expiry at a large enough value to prevent this condition, but
+<2> A `SmartLifecycle` attribute to specify whether this endpoint should start automatically after the application context start or not.
+It defaults to `true`. if `false`, RedisStreamMessageProducer should be started manually `messageProducer.start()`.
+<3> if `false`, Received messages are not auto acknowledged. The acknowledgement of the message will be deferred to the client consuming message.
+It defaults to `true`. If there is `SerializationException` message will be sent to `error-channel` and auto-acknowledged.
```
Probably this one needs a separate paragraph.
It confuses.
 You say it is auto-ack'ed. In both cases or just when `setAutoAck` is true?

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -791,12 +791,104 @@ You should set the expiry at a large enough value to prevent this condition, but
+        messageProducer.setConsumerName(""); <6>
+        messageProducer.setOutputChannel(fluxMessageChannel()); <7>
+        messageProducer.setReadOffset(""); <8>
+        messageProducer.extractPayload(""); <9>
```
Hm. I believe it was a copy/paste artefact, but it is wrong direction.
Please, revise values for the options: many of them are `boolean`, but you insert empty string.

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -791,12 +791,104 @@ You should set the expiry at a large enough value to prevent this condition, but
+It defaults to `true`. if `false`, RedisStreamMessageProducer should be started manually `messageProducer.start()`.
+<3> if `false`, Received messages are not auto acknowledged. The acknowledgement of the message will be deferred to the client consuming message.
+It defaults to `true`. If there is `SerializationException` message will be sent to `error-channel` and auto-acknowledged.
+<4> if `true`, Consumer Group will be created. During creation of consumer group stream will be created too.
```
```suggestion
<4> if `true`, Consumer Group will be created. During creation of consumer group stream will be created (if not exists yet), too.
```

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -791,12 +791,104 @@ You should set the expiry at a large enough value to prevent this condition, but
+It defaults to `true`. If there is `SerializationException` message will be sent to `error-channel` and auto-acknowledged.
+<4> if `true`, Consumer Group will be created. During creation of consumer group stream will be created too.
+It defaults to `false`.
+<5> Set Consumer Group name.
```
Why don't you explain that by default consumer group is based on the bean name?
See `ReactiveRedisStreamMessageProducer` JavaDocs.

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -791,12 +791,104 @@ You should set the expiry at a large enough value to prevent this condition, but
+<4> if `true`, Consumer Group will be created. During creation of consumer group stream will be created too.
+It defaults to `false`.
+<5> Set Consumer Group name.
+<6> Set Consumer name.
```
We must explain `consumer` and `group` relationship.

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -791,12 +791,104 @@ You should set the expiry at a large enough value to prevent this condition, but
+It defaults to `false`.
+<5> Set Consumer Group name.
+<6> Set Consumer name.
+<7> The message channel to which to send `Message` instances from this endpoint. The channel should be `FluxMessageChannel`. See <<./channel.adoc#flux-message-channel,`FluxMessageChannel`>> for more information.
```
Why are you so obsessed with `FluxMessageChannel`?
Can't we really use with this component some other channels?

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -791,12 +791,104 @@ You should set the expiry at a large enough value to prevent this condition, but
+<7> The message channel to which to send `Message` instances from this endpoint. The channel should be `FluxMessageChannel`. See <<./channel.adoc#flux-message-channel,`FluxMessageChannel`>> for more information.
+<8> Define the offset to read message.
+It defaults to `ReadOffset.latest()`.
+<9> If 'true', channel adapter will extract payload.
```
What does it mean?
May I as a new user of this component understand what is this story about?

artembilan reviewed (changes requested):
Please, find some review.

Let me know if you have time to fix it today!
Otherwise I must take it from you and fix myself.

Thanks

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -791,12 +791,104 @@ You should set the expiry at a large enough value to prevent this condition, but
-TBD
+Spring Integration 5.4 introduced the Reactive Stream inbound channel adapter for reading messages from a Redis Stream.
+Inbound channel adapter uses `StreamReceiver.receive(...)` or `StreamReceiver.receiveAutoAck` based on
+auto acknowledgement flag to read `Message` from redis stream.
```
Forgot to point to this one as well: https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line

artembilan committed:
GH-3301: Improve JMS receiveTimeout docs

Fixes https://github.com/spring-projects/spring-integration/issues/3301

Explain in docs how to configure a `JmsTemplate.receiveTimeout(-1)`

artembilan committed:
Add Java DSL for R2DBC module (#3415)

* Add Java DSL for R2DBC module

* Add JavaDocs for R2DBC component setters
* Introduce a `R2dbcMessageSource.SelectCreator` to narrow
a `StatementMapper` API access from SpEL
* Add missed `package-info.java` into R2DBC packages
* Add `hamcrest-core` as test dependency into SI-R2DBC
since Awaitility doesn't work without it

* * Add `@NonNullApi` to R2DBC packages

artembilan committed:
* Add `@NonNullApi` to ZeroMQ packages

Commit:
Fix review comments

Commit:
Fix review comments: Remove specific channel from inbound channel adapter

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -769,6 +769,94 @@ IMPORTANT: The `task-executor` has to be configured with more than one thread fo
+}
+----
+<1> Construct an instance of `ReactiveRedisStreamMessageHandler` using `ReactiveRedisConnectionFactory` and expression for `StreamKey`.
+<2> Set `RedisSerializationContext` which provides SerializationPairs for key, value, hash-key (field), hash-value and String serialization and deserialization.
```
Looks like copy/paste of the `RedisSerializationContext` JavaDocs.
What is better to see here is the purpose of this option in the `ReactiveRedisStreamMessageHandler`: what does it do for end-user data and so on.

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -769,6 +769,94 @@ IMPORTANT: The `task-executor` has to be configured with more than one thread fo
+=== Redis Stream Inbound Channel Adapter
+
+Spring Integration 5.4 introduced the Reactive Stream inbound channel adapter for reading messages from a Redis Stream.
+Inbound channel adapter uses `StreamReceiver.receive(...)` or `StreamReceiver.receiveAutoAck` based on auto acknowledgement flag to read `Message` from redis stream.
```
Yeah... It doesn't read a `Message`, but rather `Record` and then build a `Message` with an appropriate headers.

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -769,6 +769,94 @@ IMPORTANT: The `task-executor` has to be configured with more than one thread fo
+[source,java]
+----
+@Component
+public class MyClass {
```
As I said before: we don't need the whole class for configuration.
A code snippet starting with `@ServiceActivator` is fully enough in this context.

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -769,6 +769,94 @@ IMPORTANT: The `task-executor` has to be configured with more than one thread fo
+        messageProducer.setCreateConsumerGroup(true); <5>
+        messageProducer.setConsumerGroup("my-group"); <6>
+        messageProducer.setConsumerName("my-consumer"); <7>
+        messageProducer.setOutputChannel(""); <8>
```
`setOutputChannel()` accepts a `MessageChannel`.
You can use here just a plain reference: `messageProducer.setOutputChannel(fromRedisStreamChannel); <8>`

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -769,6 +769,94 @@ IMPORTANT: The `task-executor` has to be configured with more than one thread fo
+}
+----
+<1> Construct an instance of `ReactiveRedisStreamMessageProducer` using `ReactiveRedisConnectionFactory` and `StreamKey`.
+<2> `StreamReceiver.StreamReceiverOptions` Options to consume redis stream using reactive infrastructure.
```
The "Redis" word must be capitalized.

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -769,6 +769,94 @@ IMPORTANT: The `task-executor` has to be configured with more than one thread fo
+<3> A `SmartLifecycle` attribute to specify whether this endpoint should start automatically after the application context start or not.
+It defaults to `true`.
+If `false`, `RedisStreamMessageProducer` should be started manually `messageProducer.start()`.
+<4> If `false`, Received messages are not auto acknowledged. The acknowledgement of the message will be deferred to the client consuming message.
```
Why "Received is capitalized?
One sentence per line, please: https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line

The "acknowledgement" deserves a separate paragraph to explain the `IntegrationMessageHeaderAccessor.ACKNOWLEDGMENT_CALLBACK` header and how to use it.

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -769,6 +769,94 @@ IMPORTANT: The `task-executor` has to be configured with more than one thread fo
+<8> The message channel to which to send `Message` instances from this endpoint.
+<9> Define the offset to read message.
+It defaults to `ReadOffset.latest()`.
+<10> If 'true', channel adapter will extract payload from `Message`
```
That's something misleading. What does it extract and from what?
The point of this inbound endpoint is really read from Redis, build a message and produce it.
See the behavior around that `extractPayload` in the `ReactiveRedisStreamMessageProducer`

artembilan reviewed (changes requested):
You haven't rebased, but rather merged to master.
I'm not sure what my commits are doing in your PR...

Sorry for so nit-pick review, but we do here Open Source and community asks very often to give them good docs with good samples and behavior explanation.

You did great for a first time, but I need to release today. So, I take this from you and improve locally to merge the final version.

Thank you for your effort!

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/51ebf40a5f48f0ca99e828ae6985aacc50399160 after some clean up and additions.
